### PR TITLE
Assignment3_AndreeHadi

### PR DESCRIPTION
**Pull request by HadiM2**
@HadiM2 @aramirez61

Did the extra credit for:
Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)?

### CODE REVIEW

**Comment by NathanMarsee** [comment]
I'll review this

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-
+    <TargetFramework>net8.0</TargetFramework>
+    <LangVersion>latest</LangVersion>
     <IsPackable>false</IsPackable>
   </PropertyGroup>
```
Comment (NathanMarsee) [inline CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:7]: You'll probably want to copy the contents of your CanHazFunny.csproj file to this one as well.

File: CanHazFunny/CanHazFunny/IJokeOutput.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (NathanMarsee) [inline CanHazFunny/CanHazFunny/IJokeOutput.cs:5]: Minor nitpick, but most of these imports are unused. At least in my editor, this is flagged with a warning.

File: CanHazFunny/CanHazFunny/JokeOutput.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using System;
+namespace CanHazFunny;
+
+    public class JokeOutput : IJokeOutput
+    {
+        public void PrintingJoke(string jokePrint)
```
Comment (NathanMarsee) [inline CanHazFunny/CanHazFunny/JokeOutput.cs:6]: Might want to null check this, but it's not that important.

File: CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    //[inlineData("JOKE1")]
+    //[inlineData ("JOKE2")]
+    //public void PrintingJoke_OutputToConsole_Success(string output)
+    //{
+    //    JokeOutput jokePrint = new();
+    //    jokePrint.PrintingJoke(output);
+
+    //    //Mock mockConsole = new < IConsoleWrapper > ();
+    //    //string outputString =  mockConsole.WriteLine();
+
```
Comment (NathanMarsee) [inline CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs:25]: Look into using a StringWriter to change where JokeOutput prints, then you can compare to that.

**Review by NathanMarsee — APPROVED** [review]
Assignment Checklist:

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌ (70.8% line coverage as of this review)

Be sure you enable:
  Nullability for all projects ❌✔ (Not in tests .csproj configuration)
  Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
  Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌✔ (Not in tests .csproj configuration)
  Ensure that you turn on treat warnings as errors for all projects ❌✔ (Not in tests .csproj configuration)
  Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌✔ (Not in tests .csproj configuration)
  Ensure that you turn on EnforceCodeStyleInBuild ❌✔ (Not in tests .csproj configuration)
For this assignment, always use Assert.Equal<T>() (the generic version) ✔ (No Assert.Equal statements used)
Ensure there are no errors or warnings (including code analysis warnings) ❌ (Unused imports in IJokeOutput)
All of the above should be unit tested ❌(Still needs to test JokeOutput)
Choose simplicity over complexity ✔

Reviewer note:
Great work overall, just a couple of small things to change and all these ❌s should vanish

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj (hunk 1 above)
Comment (HadiM2) [reply CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj]: Thank you, Nathan, should be all good now!

File: CanHazFunny/CanHazFunny/JokeOutput.cs (hunk 3 above)
Comment (HadiM2) [reply CanHazFunny/CanHazFunny/JokeOutput.cs]: Just added a test and added the null check!

File: CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs (hunk 4 above)
Comment (HadiM2) [reply CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs:26]: Thank you for the suggestion!

File: CanHazFunny/CanHazFunny/IJokeOutput.cs (hunk 2 above)
Comment (Aramirez61) [reply CanHazFunny/CanHazFunny/IJokeOutput.cs]: Good catch, gets rid of code bloat that doesn't need to be there.

File: CanHazFunny/CanHazFunny/JokeOutput.cs (hunk 3 above)
Comment (Aramirez61) [reply CanHazFunny/CanHazFunny/JokeOutput.cs]: Gives us better code coverage, thanks.

File: .DS_Store
Comment (BenjaminMichaelis) [inline .DS_Store:1]: These should be in the `gitignore` and not commited, I expect this to be the case in Assignment 4

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 5]
```diff
@@ -1,7 +1,56 @@
...
+    private readonly Mock<IJokeService> JokeServiceMock;
+    private readonly Mock<IJokeOutput> JokeOutputMock;
+    private readonly Jester _jesterInstance;
+
+    public JesterTests()
+    {
+        JokeServiceMock = new Mock<IJokeService>();
+        JokeOutputMock = new Mock<IJokeOutput>();
+        _jesterInstance = new Jester(JokeServiceMock.Object, JokeOutputMock.Object);
+    }
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:18]: While this isn't bad in this case, generally you should put this setup in a Setup method and not in a constructor, maybe with each being an out parameter.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 6]
```diff
@@ -1,7 +1,56 @@
...
+
+
+    [Fact]
+    public void Constructor_NullJokeService_ThrowsArgumentNullException()
+    {
+        IJokeService? jokeService = null;  // Use nullable reference type annotation
+        IJokeOutput jokeOutput = Mock.Of<IJokeOutput>();
+
+        Assert.Throws<ArgumentNullException>(() => new Jester(jokeService!, jokeOutput));  // Use the null-forgiving operator (!)
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:27]:
```suggestion
        IJokeOutput jokeOutput = Mock.Of<IJokeOutput>();

        Assert.Throws<ArgumentNullException>(() => new Jester(null!, jokeOutput));  // Use the null-forgiving operator (!)

```
Can be simplified

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 7]
```diff
@@ -1,7 +1,56 @@
...
+    }
+
+    [Fact]
+    public void Constructor_NullJokeOutput_ThrowsArgumentNullException()
+    {
+        IJokeService jokeService = Mock.Of<IJokeService>();
+        IJokeOutput? jokeOutput = null;  // Use nullable reference type annotation
+
+        Assert.Throws<ArgumentNullException>(() => new Jester(jokeService, jokeOutput!));  // Use the null-forgiving operator (!)
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:36]:
```suggestion
        IJokeService jokeService = Mock.Of<IJokeService>();

        Assert.Throws<ArgumentNullException>(() => new Jester(jokeService, null!));  // Use the null-forgiving operator (!)

```
Can be simplified

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 8]
```diff
@@ -1,7 +1,56 @@
...
+
+    [Fact]
+    public void TellJoke_ChuckNorrisJoke_Skips()
+    {
+        string chuckJoke = "Chuck Norris";
+        string joke = "joke";
+        int count = 0;
+
+        JokeServiceMock?.Setup(js => js.GetJoke()).Returns(() => count++ == 0 ? chuckJoke : joke);
+
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:47]:
```suggestion
        JokeServiceMock?.SetupSequence(js => js.GetJoke()).Returns("Chuck Norris).Returns("joke");

```
The current way is a conditional and is more difficult than it needs to be. `StartupSequence` is easier

File: CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            JokeOutput output = new();
+            using StringWriter sw = new();
+            Console.SetOut(sw);
+            output.PrintingJoke(null!);
+        }
+        catch (ArgumentNullException)
+        {
+            exceptionThrown = true;
+        }
+        Assert.True(exceptionThrown);
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny.Tests/JokeOutputTests.cs:43]: Way more difficult than it needs to be. Using `Assert.Throws<ArgumentNullException>` is the proper way to do this check

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj [hunk 10]
```diff
@@ -1,8 +1,15 @@
...
-</Project>
+<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<OutputType>Exe</OutputType>
+		<TargetFramework>net8.0</TargetFramework>
+		<LangVersion>latest</LangVersion>
+		<Nullable>enable</Nullable>
+		<EnableNETAnalyzers>true</EnableNETAnalyzers>
+		<Nullability>true</Nullability>
```
Comment (BenjaminMichaelis) [inline CanHazFunny/CanHazFunny/CanHazFunny.csproj:9]: Not needed

**Pull request closed without merging**